symbolic-base/src/ZkFold/Base/Protocol/Protostar/AccumulatorScheme.hs

```diff
@@ -22,94 +23,75 @@ import           ZkFold.Base.Protocol.Protostar.CommitOpen   (CommitOpen (..), C
-           -> [c]                        -- NARK proof π.x
-           -> AccumulatorInstance i f c  -- accumulator instance acc.x
-           -> AccumulatorInstance i f c  -- updated accumulator instance acc'.x
-           -> [c]                        -- accumulation proof E_j
-           -> Bool ctx
+           -> f                           -- Commitment key ck
+           -> Accumulator pi f c m        -- accumulator
+           -> InstanceProofPair pi c m    -- instance-proof pair (pi, π)
+           -> (Accumulator pi f c m, [c]) -- updated accumulator and accumulation proof
+
+  verifier :: pi                          -- Public input
+           -> [c]                         -- NARK proof π.x
+           -> AccumulatorInstance pi f c  -- accumulator instance acc.x
+           -> AccumulatorInstance pi f c  -- updated accumulator instance acc'.x
+           -> [c]                         -- accumulation proof E_j
+           -> (f, pi, [f], [c], c)        -- returns zeros if the accumulation proof is correct
 
   decider  :: a
-           -> (f, KeyScale f)       -- Commitment key ck and scaling factor
-           -> Accumulator i f c m   -- final accumulator
-           -> Bool ctx
-
-data KeyScale f = KeyScale f f
-    deriving (P.Show, Generic, NFData)
-
--- | Class describing types which can form a polynomial linear combination:
--- linearCombination a1 a2 -> a1 * X + a2
---
-class LinearCombination a b where
-    linearCombination :: a -> a -> b
-
--- | Same as above, but with a coefficient known at runtime
--- linearCombination coeff b1 b2 -> b1 * coeff + b2
---
-class LinearCombinationWith a b where
-    linearCombinationWith :: a -> b -> b -> b
+           -> f                           -- Commitment key ck
+           -> Accumulator pi f c m        -- final accumulator
+           -> ([c], c)                    -- returns zeros if the final accumulator is valid
 
-instance (Scale f a, AdditiveSemigroup a) => LinearCombinationWith f [a] where
-    linearCombinationWith f = P.zipWith (\a b -> scale f a + b)
+type DataRepresentableAsList f x = (IsList x, Item x ~ f)
 
 instance
-    ( Symbolic ctx
-    , Eq (Bool ctx) c
-    , Eq (Bool ctx) i
-    , Eq (Bool ctx) f
-    , Eq (Bool ctx) [f]
-    , Eq (Bool ctx) [c]
+    ( AdditiveGroup pi
     , AdditiveGroup c
     , AdditiveSemigroup m
     , Ring f
+    , Scale f pi
     , Scale f c
     , Scale f m
-    , MapInput f a ~ i
-    , deg ~ Degree (CommitOpen m c a) + 1
-    , KnownNat deg
-    , LinearCombination (MapMessage f a) (MapMessage (PU.PolyVec f deg) a)
-    , LinearCombination (MapInput f a) (MapInput (PU.PolyVec f deg) a)
-    , LinearCombinationWith f (MapInput f a)
-    , MapMessage f a ~ m
+    , RandomOracle pi f         -- Random oracle for compressing public input
+    , RandomOracle c f          -- Random oracle ρ_NARK
+    , HomomorphicCommit f [f] c
+    , HomomorphicCommit f m c
+    , IsList pi
+    , Item pi ~ f
+    , IsList m
+    , Item m ~ f
```

Comment:
Is this true, though? Prover messages in the Special sound protocol for circuits are Maps. Moreover, they require a different approach to linear combinations than lists: absent keys are treated as zero coefficients. So, `linearCombination [('a', 1), ('c', 3)] [('b', 2), ('c', 4)]` should yield `[('a', X), ('b', 2), ('c', 3X + 4)]`. That's why I introduced the `LinearCombination` class

Reply:
Still, I think `LinearCombination a b` can be replaced with `Scale a b, AdditiveSemigroup b`

Reply:
> Is this true, though? Prover messages in the Special sound protocol for circuits are Maps. Moreover, they require a different approach to linear combinations than lists: absent keys are treated as zero coefficients. So, `linearCombination [('a', 1), ('c', 3)] [('b', 2), ('c', 4)]` should yield `[('a', X), ('b', 2), ('c', 3X + 4)]`. That's why I introduced the `LinearCombination` class

It does not matter the context. In my understanding of the paper, the assumption is that `pi` and `m` can be somehow broken down into field elements (of the same context). So, I think `IsList` fits very well here.

Reply:
For example, `pi` could be some tuple or vector of field elements. Obviously, there is a way to write it as a list and that's all we need. We just need to make sure that it implements `toList`.

Reply:
For `pi` this is not explicitly stated. But for messages, see page 39. Clearly, all messages are just vectors of `f` of the appropriate size. When we apply the "commit-open" and "Fiat-Shamir" transformations we only use tuple and list operators, so `toList` can still be easily defined for such messages. Actually, if we promote the number of rounds to the type level, then we only use the tuple operator.